torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +856,335 @@ def _get_ignored_params(
+        function is not ``None`` and falls back to default behavior otherwise.
+        For meta device, this calls ``reset_parameters()``, and for torchdistX
+        fake tensors, this calls ``deferred_init.materialize_module()``.
+
+        Precondition: ``self.compute_device`` is set via
+        ``_get_compute_device()``.
+        """
+        is_meta_module = any(p.is_meta for p in module.parameters())
+        is_torchdistX_deferred_init = (
+            not is_meta_module
+            and _TORCHDISTX_AVAIL
+            and any(fake.is_fake(p) for p in module.parameters())
+        )
+        if (
+            is_meta_module or is_torchdistX_deferred_init
+        ) and param_init_fn is not None:
+            if not callable(param_init_fn):
+                raise ValueError(
+                    f"Expected {param_init_fn} to be callable but got {type(param_init_fn)}"
+                )
+            param_init_fn(module)
+        elif is_meta_module:
+            # Run default meta device initialization
+            module.to_empty(device=self.compute_device)
+            try:
+                with torch.no_grad():
+                    module.reset_parameters()
+            except BaseException as e:
+                warnings.warn(
+                    "Unable to call `reset_parameters()` for module on meta "
+                    f"device with error {str(e)}. Please ensure your "
+                    "module implements a `reset_parameters()` method."
+                )
+                raise e
+        elif is_torchdistX_deferred_init:
+            # Run default torchdistX initialization
+            deferred_init.materialize_module(
+                module,
+                check_fn=lambda k: not isinstance(k, FullyShardedDataParallel),
+            )
+
+    def _move_module_to_device(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+        device_id: Optional[Union[int, torch.device]],
+    ):
+        """
+        Moves ``module`` depending on ``device_id`` and its current device,
+        where ``device_id`` should be what the user passed into the FSDP
+        constructor. Since the entire module is moved, ignored parameters are
+        moved as well.
+
+        - If ``device_id`` is not ``None``, then this moves ``module`` to the
+        corresponding device.
+        - If ``device_id`` is ``None``, then this does not move ``module`` but
+        warns the user if it is only CPU.
+
+        Precondition: ``_check_single_device_module()``.
+        """
+        cpu_device = torch.device("cpu")
+        param = next(self._get_orig_params(module, ignored_params), None)
+        if param is None:
+            return  # no original parameters to manage
+        if device_id is not None:
+            assert param.device == cpu_device or param.device == self.compute_device, (
+                f"Unexpected device {param.device} -- check the FSDP implementation"
```

Comment:
maybe change this to "please file an issue with a repro" since it indicates a development bug